cpp/ql/lib/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll

```diff
@@ -688,6 +688,11 @@ private Overlap getExtentOverlap(MemoryLocation0 def, MemoryLocation0 use) {
     (
       // EntireAllocationMemoryLocation exactly overlaps itself.
       use instanceof EntireAllocationMemoryLocation and
+      (
+        def.getAnAllocation() = use.getAnAllocation() or
+        not exists(def.getAnAllocation()) or
+        not exists(def.getAnAllocation())
```

Comment:
And finally, if both `def` and `use` are `EntireAllocationMemoryLocation`s and they have the same virtual variable (as they do in this case), I think `def.getAnAllocation() = use.getAnAllocation()` always holds. So if this speeds up anything I'm pretty sure it's simply because you're permuting some joins?

Reply:
> I guess you meant for this to be `use.getAnAllocation()`?

Yes.  I wasn't 100% sure these lines were required, but based on nearby cases I thought they might be.  You seem to think not, so I'll delete them.

> I think def.getAnAllocation() = use.getAnAllocation() always holds.

I'll check again, but I'm fairly sure I had found a case where it didn't (and that fact exploded performance).  That might imply there's a deeper bug to be fixed and this change would only patch it.  I'll investigate further...

Reply:
> And finally, if both def and use are EntireAllocationMemoryLocations and they have the same virtual variable (as they do in this case), I think def.getAnAllocation() = use.getAnAllocation() always holds. 

This has been a bit of a pain to investigate, but it's definitely not the case on the `AcademySoftwareFoundation/openexr` snapshot or the MRVA top 10.  There are thousands of exceptions (possibly many more, I had to arbitrarily constrain the query so it would execute in reasonable time).  Many times the issue seems to occur when a [template?] function takes multiple pointer arguments, between the memory locations of those arguments.  Other cases seem to relate to dynamic allocations.

The question I'm trying to figure out next is why these cases occur - and whether they should be fixed at the source, or if we in fact should not be assuming `def.getAnAllocation() = use.getAnAllocation()` holds here.

Reply:
I still have a limited understanding of how this works, but it seems to me that if two or more allocations in the same function escape, they end up in the same virtual variable (`TAllAliasedMemory` for the function) but with different allocations (`EntireAllocationMemoryLocation` being smaller than `TAllAliasedMemory`).  So from the point of view of `getExtentOverlap`, I think it's right that we have to check that `def.getAnAllocation() = use.getAnAllocation()` for `EntireAllocationMemoryLocation`s.

So I tentatively think my change here is correct.  And I've removed the two lines you pointed out are redundant now.  I think the next step is to *try* to come up with a test case where the difference matters.

Reply:
OK, so we're happy that `def.getAnAllocation() = use.getAnAllocation()` should affect results in theory, it's more than just a performance hint.

The code you wrote in the comment immediately above is very similar to some of the real world cases I've looked at.  I've not been able to find an existing test that exposes this behaviour though (perhaps I need to write a new one?)

Reply:
@MathiasVP do you have any idea how best to expose the behaviour of this predicate in tests?  Short of writing a direct test for `getExtentOverlap` itself, which I suppose is an option.

Reply:
A value numbering test is probably the easiest way to get hard proofs of the change in behavior since:
- That's a library that's directly used in queries, and
- That's a library that makes full use of alias analysis